### PR TITLE
Docs: Add introduction documentation for cairo vm - Issue #328

### PR DESCRIPTION
**Pull request by coxmars**
This PR is for the issue Intro Documentation #328. Added intro documentation to understand in a general way all the sections as a small summary of what we'll see in the specific parts.

### CODE REVIEW

**Review by TAdev0 — CHANGES_REQUESTED** [review]
Hey @coxmars 

I left a first bunch of comments on your PR but it looks good overall!! 

Did you use AI for the part about builtins? We should avoid  things like: 

" These include managing resource usage effectively to prevent performance degradation or vulnerabilities, ensuring robust security measures to mitigate risks such as integer overflow, reentrancy attacks, and access control issues, and maintaining compatibility with existing standards and protocols within the blockchain ecosystem. "

**Comment by coxmars** [comment]
> Hey @coxmars
> 
> I left a first bunch of comments on your PR but it looks good overall!!
> 
> Did you use AI for the part about builtins? We should avoid things like:
> 
> " These include managing resource usage effectively to prevent performance degradation or vulnerabilities, ensuring robust security measures to mitigate risks such as integer overflow, reentrancy attacks, and access control issues, and maintaining compatibility with existing standards and protocols within the blockchain ecosystem. "

Thank you for the feedback, I will apply the changes as soon as possible and I will modify that sentence to avoid that kind of explanation

**Comment by coxmars** [comment]
Hi @TAdev0 can you review these changes please, I considered all your comments for this improvement but if something is needed to improve I will do it, thanks.

File: docs/docs/vm-fundamentals/intro.md [hunk 1]
```diff
@@ -2,6 +2,73 @@
...
+
+## Resources
+
+In the Cairo Virtual Machine, state transitions, influenced by each instruction, play a crucial role in the VM's operation. Unlike traditional deterministic CPU architectures, Cairo can handle nondeterministic transitions, allowing it to verify a wider range of computational statements efficiently. However, if a valid subsequent state cannot be established leading to what's known as an "execution rejection" the prover cannot generate proof for that program's execution. Therefore, this highlights the importance of stringent resource management within Cairo, ensuring that the VM operates within its limits and maintains the overall stability and security of the Starknet platform.
+
+Managing resources effectively is essential for preventing scenarios where an execution might be rejected due to the system hitting resource limits or encountering computationally infeasible states. By carefully monitoring and constraining the use of memory, computational steps, and data outputs, the Cairo VM ensures that each program can be executed and verified without risking the overall stability and security of the Starknet platform.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:74]: All this is not accurate.

The CASM bytecode by construction will never fail. It is not possible to compile cairo 0 or cairo 1 code that can fail

See
https://medium.com/nethermind-eth/under-the-hood-of-cairo-1-0-exploring-sierra-7f32808421f5

This means the Cairo VM is just here tu run CASM bytecode that we know will not fail.

> By carefully monitoring and constraining the use of memory, computational steps, and data outputs, the Cairo VM ensures that each program can be executed and verified without risking the overall stability and security of the Starknet platform.

The cairo VM just executes Casm bytecode,

**Review by TAdev0 — CHANGES_REQUESTED** [review]
Left another bunch of comments!

**Comment by TAdev0** [comment]
hi @coxmars  any update on this?

**Comment by coxmars** [comment]
> hi @coxmars any update on this?

I'm finishing applying the changes, I'll let you know as soon as I have it.

**Comment by rodrigo-pino** [comment]
Hey @coxmars if you need any help please let us know

**Comment by coxmars** [comment]
> Hey @coxmars if you need any help please let us know

Hi Rodrigo, I would like you to give me some context/guidance on the resources section, what do you mean by that? I did not find related information or maybe I did not understand the specific approach to address the explanation. Thank you.

File: docs/package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "@docusaurus/preset-classic": "2.4.1",
     "@mdx-js/react": "^1.6.22",
     "clsx": "^1.2.1",
+    "docusaurus": "^1.14.7",
```
Comment (rodrigo-pino) [inline docs/package.json:21]: You don't need to add this dependency for docsaurus to work, could you please remove it

File: docs/package-lock.json
Comment (rodrigo-pino) [inline docs/package-lock.json:1]: Which also measn this file doesn't need to be changed

File: docs/package.json (hunk 2 above)
Comment (coxmars) [reply docs/package.json]: Done

File: docs/package-lock.json
Comment (coxmars) [reply docs/package-lock.json:1]: Done

File: docs/docs/vm-fundamentals/intro.md [hunk 3]
```diff
@@ -2,6 +2,86 @@
...
+# Introduction to the Cairo Virtual Machine
 
-An introduction to fundamentals. Which core concepts we will be seeing, what are their purpose etc. Maybe a little description of each one.
+Welcome to the documentation for the Cairo VM. 
+
+First, let us clarify what is a Virtual Machine (VM)? A VM is **a digital version of a physical computer,** that can run programs and operating systems, store data, connect to networks, and do other computing functions.
+
+This documentation is designed to provide developers with a fundamental understanding of the Cairo VM's architecture and capabilities, empowering them to build and optimize applications on Starknet effectively. This documentation will be divided into four main parts, each of which we will briefly describe in this introduction.
+
+The Cairo Virtual Machine (CVM), at the heart of the Starknet network, is an advanced computational framework specifically designed to execute Casm bytecode generated by the Cairo compiler. Also, the Cairo VM is able to generate cryptographic proofs, these proofs are be posted on a settlement layer such as Ethereum, offering Starknet the same level of security as the underlying layer
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:13]: > are be posted

did you mean `will be posted`?

Comment (TAdev0) [reply docs/docs/vm-fundamentals/intro.md:13]: missing point at the end of the last sentence

File: docs/docs/vm-fundamentals/intro.md [hunk 4]
```diff
@@ -2,6 +2,86 @@
...
+
+## VM Layout
+
+The Cairo Virtual Machine optimizes its operations around a unique structure of registers and memory to facilitate the efficient execution of Casm bytecode. 
+
+### Registers
+
+Unlike traditional architectures where registers are separate storage units, in Cairo, these registers are pointers that facilitate access and management of memory during program execution.
+
+General-Purpose Registers: These are hypothetical in Cairo, where instead of traditional CPU-bound registers all operations are directly performed on memory cells. This approach aligns with Cairo’s optimization strategy which aims to minimize the number of trace cells in the Algebraic Intermediate Representation (AIR). In Cairo, registers such as the Program Counter (pc), Allocation Pointer (ap), and Frame Pointer (fp) do not store data themselves, instead:
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:23]: end with a point before the next section. You can remove instead: i think

File: docs/docs/vm-fundamentals/intro.md [hunk 5]
```diff
@@ -2,6 +2,86 @@
...
+
+### Pointers
+
+Program Counter (pc) points to the current instruction in memory that the VM is executing.
+Allocation Pointer (ap) identifies the first free memory cell not yet used by the program, crucial for dynamic memory management. Frame Pointer (fp) points to the beginning of the stack frame of the current function, managing the execution context of function calls.
+
+### Memory
+
+Cairo employs a nondeterministic read-only memory model. Unlike typical random-access memory (RAM) used in most computing systems where any memory cell can be read or written at any time, Cairo's memory model only allows memory to be written once and restricts subsequent modifications. In addition, this model is particularly efficient for AIR, as it requires fewer trace cells per memory access, optimizing the proof generation process.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:32]: i'm not sure `particularly efficient for AIR` is the proper way to say this.

Maybe  : `In addition, this memory model is particularly efficient for proof generation, as it requires fewer trace cells per memory access.`

File: docs/docs/vm-fundamentals/intro.md [hunk 6]
```diff
@@ -2,6 +2,86 @@
...
+
+Cairo's memory model also introduces the idea of "public memory," a crucial feature for validating data integrity between the prover and verifier. This feature enables the prover to prove that specific memory cells contain certain values without exposing the entire memory's content. 
+
+This feature is crucial for operations like loading bytecode, passing arguments, and returning values from programs because it ensures data integrity and security. By proving that specific memory cells contain certain values without exposing the entire memory's content, it prevents unauthorized access to data and protects against data manipulation. Additionally, it increases efficiency as it allows the Cairo VM to directly access required data instead of scanning through the entire memory.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:36]: i dont know if we should keep this section. I dont feel like i understand a concept when reading it. It feels like a summary of some technical content in the cairo paper

File: docs/docs/vm-fundamentals/intro.md [hunk 7]
```diff
@@ -2,6 +2,86 @@
...
+
+## Basic Instructions
+
+Cairo's instruction set is designed to facilitate the verification of computations within its unique programming environment. The foundational aspect of Cairo instructions revolves around assertions of equality, which form the basis of most operations within the virtual machine.
+
+### Assert Statements
+
+A primary form of instruction in Cairo is the assert statement, which verifies that two values or computations are equal. For example, the instruction `[ap] = [ap - 1] * [fp], ap++`; not only performs multiplication but also asserts that the result is equal to the value at the current free memory cell `[ap]`, before incrementing `[ap]` with `ap++`. This is crucial for the generation of proofs, however, proofs cannot be partially correct, they can be generated or not
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:44]: > This is crucial for the generation of proofs, however, proofs cannot be partially correct, they can be generated or not

this should be improved, or just remove it for now

File: docs/docs/vm-fundamentals/intro.md [hunk 8]
```diff
@@ -2,6 +2,86 @@
...
+
+### Pointer Management
+
+The incrementing of the allocation pointer (ap++) is an important standalone operation in many Cairo instructions. It readies the virtual machine to utilize the next cell in memory for future operations, thereby ensuring a seamless flow of computation.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:48]: > incrementing

increment?

> readies 

prepares?

File: docs/docs/vm-fundamentals/intro.md [hunk 9]
```diff
@@ -2,6 +2,86 @@
...
+
+### Advanced Syntax and Operations
+
+In addition to fundamental arithmetic and control flow operations, Cairo provides advanced features for greater code complexity and efficiency. It supports complex operations such as double dereferencing, which allows values to be used as memory addresses, and the direct modification of values using immediate values and offsets, which provides flexibility in memory operations.
+
+## Builtins
+
+Builtins in Cairo are predefined and optimized low-level execution units that are added to the Cairo CPU board to perform predefined calculations that are expensive to perform in Cairo. Communication between the CPU and embedded functionalities occurs through memory-mapped I/O. Each builtin is assigned a contiguous region of memory and applies specific restrictions on the data found in that area.
+
+This builtins are: output, Pedersen, range_check, ecdsa and bitwise. Each of this would be onboard in the specific section for builtins in this documentation.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:67]: these are cairo zero builtins. There are others. You can remove it for now

File: docs/docs/vm-fundamentals/intro.md [hunk 10]
```diff
@@ -2,6 +2,86 @@
...
+
+A "hint" in Cairo is a block of any language executed just before the subsequent Cairo instruction, interacting directly with the program’s variables or memory. To incorporate a hint, simply write Python code inside the `%{ ... %}` markers, this feature allows programmers to leverage Python's extensive functionalities within Cairo programs, enhancing their capability to manage complex computations efficiently and securely.
+
+Hints in Cairo are useful when the cost of verifying a computed value is cheaper than finding the value itself, they are particularly valuable for solving equations or optimizing computational tasks. When considering hints, you should evaluate the properties of the inputs you will provide, the inequalities that define them, and the tools and builtins already available in Cairo.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:77]: I think you can remove the last part for now which is a bit  unclear. hints are mostly used under the hood, and devs dont write hints. But its possible to use them in Cairo programs indeed, to achieve privacy for example.

File: docs/docs/vm-fundamentals/intro.md [hunk 11]
```diff
@@ -2,6 +2,86 @@
...
+
+## Cairo Runner
+
+The Cairo Runner, a specialized computer program, executes compiled Cairo programs. Unlike conventional programs, these include nondeterministic code due to Cairo's acceptance of nondeterminism. For instance, when computing the square root of a number, the Cairo Runner considers both possible outcomes (+Y and -Y). It then selects the solution that aligns with the following program instructions.
+
+The output of the Cairo Runner comprises two main parts: an accepting input for the Cairo nondeterministic machine and an accepting input for the Cairo deterministic machine, which serves as the witness for the nondeterministic machine. Additionally, the Runner may return a failure if the execution results in a contradiction or if it cannot compute the value of a memory cell due to insufficient hints.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:83]: > an accepting input for the Cairo nondeterministic machine and an accepting input for the Cairo deterministic machine

repetition

File: docs/docs/vm-fundamentals/intro.md (hunk 6 above)
Comment (coxmars) [reply docs/docs/vm-fundamentals/intro.md:36]: I think it could be kept for the moment to get an idea, and then can be modified or eliminated

**Comment by coxmars** [comment]
@TAdev0 improvements applied if you need other changes just let me know, thanks

File: docs/docs/vm-fundamentals/intro.md [hunk 12]
```diff
@@ -29,7 +29,7 @@ Allocation Pointer (ap) identifies the first free memory cell not yet used by th
 
 ### Memory
 
-Cairo employs a nondeterministic read-only memory model. Unlike typical random-access memory (RAM) used in most computing systems where any memory cell can be read or written at any time, Cairo's memory model only allows memory to be written once and restricts subsequent modifications. In addition, this model is particularly efficient for AIR, as it requires fewer trace cells per memory access, optimizing the proof generation process.
+Cairo employs a nondeterministic read-only memory model. Unlike typical random-access memory (RAM) used in most computing systems where any memory cell can be read or written at any time, Cairo's memory model only allows memory to be written once and restricts subsequent modifications. In addition, this memory model is particularly efficient for proof generation, as it requires fewer trace cells per memory access.
```
Comment (TAdev0) [inline docs/docs/vm-fundamentals/intro.md:32]: better:) great!

**Review by TAdev0 — APPROVED** [review]

**Review by cicr99 — APPROVED** [review]

**Pull request merged**